File: 2-sat.cpp
// BOJ 11281 2-SAT - 4

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

int dfs(int cur, int &idx, int &scc_num, vector<int> &v, vector<int> &vst, vector<int> &scc_idx, vector<int> &finish, vector<vector<int>> &graph) {
    v.push_back(cur);
    vst[cur] = idx;
    int low = idx++;

    for (int nxt : graph[cur]) {
        if (!vst[nxt])
            low = min(low, dfs(nxt, idx, scc_num, v, vst, scc_idx, finish, graph));
        else if (!finish[nxt])
            low = min(low, vst[nxt]);
    }

    if (low != vst[cur])
        return low;

    while (1) {
        int top = v.bk;
        v.pop_back();

        finish[top] = 1;
        scc_idx[top] = scc_num;

        if (cur == top)
            break;
    }

    scc_num++;
    return low;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> graph(2 * n + 1);

    while (m--) {
        int x, y;
        cin >> x >> y;

        if (x > 0 && y > 0) {
            graph[2 * x].push_back(2 * y - 1);
            graph[2 * y].push_back(2 * x - 1);
        } else if (x > 0 && y < 0) {
            graph[2 * x].push_back(-2 * y);
            graph[-2 * y - 1].push_back(2 * x - 1);
        } else if (x < 0 && y > 0) {
            graph[-2 * x - 1].push_back(2 * y - 1);
            graph[2 * y].push_back(-2 * x);
        } else if (x < 0 && y < 0) {
            graph[-2 * x - 1].push_back(-2 * y);
            graph[-2 * y - 1].push_back(-2 * x);
        }
    }

    int idx = 1;
    int scc_num = 1;
    vector<int> v, vst(2 * n + 1), scc_idx(2 * n + 1), finish(2 * n + 1);

    for (int i = 1; i <= 2 * n; i++)
        if (!vst[i])
            dfs(i, idx, scc_num, v, vst, scc_idx, finish, graph);

    for (int i = 1; i <= n; i++) {
        if (scc_idx[2 * i - 1] == scc_idx[2 * i]) {
            cout << 0;
            return 0;
        }
    }

    cout << 1 << '\n';
    for (int i = 1; i <= n; i++)
        cout << (scc_idx[2 * i - 1] < scc_idx[2 * i]) << ' ';
}

File: 2-sat.py
# BOJ 11281 2-SAT - 4

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
graph = [[] for _ in range(2 * n + 1)]

for _ in range(m):
    a, b = map(int, input().split())
    graph[-a].append(b)
    graph[-b].append(a)

stack = []
scc_num = 1
scc_idx = [0] * (2 * n + 1)
finish = [0] * (2 * n + 1)
visit = [0] * (2 * n + 1)
idx = 1


def dfs(cur):
    global idx, scc_num
    visit[cur] = idx
    low = idx
    idx += 1
    stack.append(cur)

    for nxt in graph[cur]:
        if not visit[nxt]:
            low = min(low, dfs(nxt))
        elif not finish[nxt]:
            low = min(low, visit[nxt])

    if low == visit[cur]:
        while stack:
            top = stack.pop()
            finish[top] = 1
            scc_idx[top] = scc_num
            if cur == top:
                break
        scc_num += 1
    return low


for i in range(1, 2 * n + 1):
    if not visit[i]:
        dfs(i)

res = [0] * n
for i in range(1, n + 1):
    if scc_idx[i] == scc_idx[-i]:
        print(0)
        break
    if scc_idx[i] < scc_idx[-i]:
        res[i - 1] = 1
else:
    print(1)
    print(*res)

----------------------------------------------------------------------------------------------------

n, m = map(int, input().split())

graph = [[] for _ in range(2 * n + 1)]
reverse = [[] for _ in range(2 * n + 1)]

for _ in range(m):
    a, b = map(int, input().split())
    graph[-a].append(b)
    graph[-b].append(a)
    reverse[b].append(-a)
    reverse[a].append(-b)

stack = []
visit = [0] * (2 * n + 1)
for i in range(1, 2 * n + 1):
    if not visit[i]:
        visit[i] = 1
        queue = [i]
        while queue:
            cur = queue[-1]
            for nxt in graph[cur]:
                if not visit[nxt]:
                    visit[nxt] = 1
                    queue.append(nxt)
                    break
            else:
                stack.append(queue.pop())

scc_num = 0
scc_idx = [0] * (2 * n + 1)
finish = [0] * (2 * n + 1)
while stack:
    node = stack.pop()
    if not finish[node]:
        finish[node] = 1
        queue = [node]
        while queue:
            cur = queue[-1]
            for nxt in reverse[cur]:
                if not finish[nxt]:
                    finish[nxt] = 1
                    queue.append(nxt)
                    break
            else:
                scc_idx[queue.pop()] = scc_num

        scc_num += 1

res = [0] * n
for i in range(1, n + 1):
    if scc_idx[i] == scc_idx[-i]:
        print(0)
        break
    if scc_idx[i] > scc_idx[-i]:
        res[i - 1] = 1
else:
    print(1)
    print(*res)
